dll/win32/setupapi/cfgmgr.c

```diff
@@ -7489,10 +7489,10 @@ CM_Request_Device_Eject_ExA(
     TRACE("CM_Request_Device_Eject_ExA(%lx %p %s %lu %lx %p)\n",
           dnDevInst, pVetoType, debugstr_a(pszVetoName), ulNameLength, ulFlags, hMachine);
 
-    if (pszVetoName == NULL && ulNameLength == 0)
+    if (pszVetoName == NULL && ulNameLength != 0)
         return CR_INVALID_POINTER;
 
-    lpLocalVetoName = HeapAlloc(GetProcessHeap(), 0, ulNameLength * sizeof(WCHAR));
+    lpLocalVetoName = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, (ulNameLength + 1) * sizeof(WCHAR));
```

Comment:
```suggestion
    lpLocalVetoName = HeapAlloc(GetProcessHeap(), 0, (ulNameLength + 1) * sizeof(WCHAR));
    lpLocalVetoName[ulNameLength] = UNICODE_NULL;
```

Reply:
Why this suggestion?

Reply:
It makes the intent (= ensuring null-termination) clearer (and maybe cheaper).
Yeah, the assignment should be moved to after that check, or after the W call for extra safety.

Reply:
No it doesn't make much sense. Because, it's not used by the W call as input, but just as an output buffer. It is the job of the W function to correctly put a NULL-terminated string in that buffer.

Reply:
> it doesn't make much sense. Because, it's not used by the W call as input, but just as an output buffer.

This is exactly why `HEAP_ZERO_MEMORY` is confusing/overkill.

> It is the job of the W function to correctly put a NULL-terminated string in that buffer.

I think it can't, as the new/actual length is not passed to it...

Reply:
the length of the allocated buffer is passed to the W function:
```c
    ret = CM_Request_Device_Eject_ExW(dnDevInst, pVetoType, lpLocalVetoName /*** <---- The output buffer ***/,
                                      ulNameLength /*** <---- Its size ***/, ulFlags, hMachine);
```

**HOWEVER:** it should be the full size (in characters) of that buffer. Here, indeed, there seems to be a bug somehow, because the buffer is allocated as having ulNameLength + 1 characters, but only ulNameLength is passed to the W function.

Reply:
> **HOWEVER:**

Exactly! This is why that new change needs to be explained/fixed.

Reply:
> HOWEVER: it should be the full size (in characters) of that buffer.

I forgot about that. Will fix.

Reply:
Thanks.

Reply:
> HOWEVER: it should be the full size (in characters) of that buffer.

The additional length is only used to make the allocation valid in case of `ulNameLength == 0` so we don't want to pass it to the W call because we need only `ulNameLength` characters.

> Yeah, the assignment should be moved to after that check, or after the W call for extra safety.

This can be done, for readability and simplicity.

Reply:
> only used to make the allocation valid

Ah, I guessed so.
You should rather allocate/convert/free only `if (pszVetoName != NULL && ulNameLength != 0)`.

Reply:
Fixed.